### ISSUE TITLE
[Bug] corrupted file on reupload, even after seaf-fsck.sh repair

### ISSUE THREAD

**Issue by T-X** [haiwen/seafile#2201]
Hi,

With seafile v6.2.5 I'm currently experiencing the following issue:

1) A download of a file via the web-ui or seafile-applet fails
2) seaf-fsck.sh finds an error and tries to repair it:

<pre><code>$ seaf-fsck.sh --repair

Starting seaf-fsck, please wait ...

[06/09/19 00:12:26] fsck.c(586): Running fsck for repo 8884c89b-685a-4444-9725-635b7780c905.
[06/09/19 00:12:26] fsck.c(413): Checking file system integrity of repo foobar(8884c89b)...
[06/09/19 00:17:13] fsck.c(105): Block 6685f8d21805f029ab451939e9a53fa3f4ffdb23 is damaged, remove it.
[06/09/19 00:17:13] fsck.c(182): File /somefile(aa79ad1f) is damaged, recreate an empty file.
[06/09/19 00:20:16] fsck.c(391): Update repo 8884c89b status to commit 6e45c540.
[06/09/19 00:20:16] fsck.c(650): Fsck finished for repo 8884c89b.
[...]
</code></pre>

3) The file is then shown with zero size, as expected.
4) Then creating a new subdirectory in the same library and reuploading the original, undamaged file there, via the web-ui. Succeeds with no error.
5) Then, however when trying to download the file it always fails at 89%:

<pre><code>$ wget --tries=3 "https://example.com/f/49327fd9086d41b18518/?dl=1"                                                                                                                    <─(Sun,Jun09)─┘
--2019-06-09 14:44:33--  https://example.com/f/49327fd9086d41b18518/?dl=1
Resolving example.com (example.com)... 2001:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx
Connecting to example.com (example.com)|2001:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx|:443... connected.
HTTP request sent, awaiting response... 302 FOUND
Location: https://example.com/seafhttp/files/98ceee2d-bf73-48c3-86f2-8840de3db9fa/somefile [following]
--2019-06-09 14:44:33--  https://example.com/seafhttp/files/98ceee2d-bf73-48c3-86f2-8840de3db9fa/somefile
Reusing existing connection to [example.com]:443.
HTTP request sent, awaiting response... 200 OK
Length: 367097856 (350M) [application/octet-stream]
Saving to: ‘index.html?dl=1’

index.html?dl=1                                       89%[=========================================================================================================>             ] 314.09M  6.56MB/s    in 56s     

2019-06-09 14:45:29 (5.64 MB/s) - Connection closed at byte 329349120. Retrying.

--2019-06-09 14:45:30--  (try: 2)  https://example.com/seafhttp/files/98ceee2d-bf73-48c3-86f2-8840de3db9fa/somefile
Connecting to example.com (example.com)|2001:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx|:443... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 367097856 (350M), 37748736 (36M) remaining [application/octet-stream]
Saving to: ‘index.html?dl=1’

index.html?dl=1                                       89%[++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++             ] 314.09M  --.-KB/s    in 0s      

2019-06-09 14:45:30 (0.00 B/s) - Connection closed at byte 329349120. Retrying.

--2019-06-09 14:45:32--  (try: 3)  https://example.com/seafhttp/files/98ceee2d-bf73-48c3-86f2-8840de3db9fa/somefile
Connecting to example.com (example.com)|2001:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx|:443... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 367097856 (350M), 37748736 (36M) remaining [application/octet-stream]
Saving to: ‘index.html?dl=1’

index.html?dl=1                                       89%[++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++             ] 314.09M  --.-KB/s    in 0s      

2019-06-09 14:45:32 (0.00 B/s) - Connection closed at byte 329349120. Giving up.
</code></pre>

The original file has a size of 367097856 bytes (which matches the "Length:" attribute above).

Rerunning seaf-fsck.sh just brings me back to step two.

**Comment by killing** [comment]
Can you check the block object in your file system? The path should be seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23

**Comment by T-X** [comment]
Hi @killing, thanks for the quick reply!

Yes, it exists:

<pre><code>seafile@seafile:~$ stat ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23 
  File: /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
  Size: 4194304         Blocks: 8192       IO Block: 4096   regular file
Device: fe1dh/65053d    Inode: 198149177   Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/ seafile)   Gid: ( 1000/ seafile)
Access: 2019-06-12 12:21:46.886628204 +0000
Modify: 2019-06-09 11:46:12.709457501 +0000
Change: 2019-06-09 11:46:12.753458582 +0000
 Birth: -
seafile@seafile:~$ cat ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23 > /dev/null; echo $?
0
</code></pre>

So reading from it seems to work fine. And seems to be exactly 4MiB as it is supposed to be?

Anything else I should check about it?

**Comment by T-X** [comment]
Also, some additional information: This is a RAID6 with 5 HDDs connected to a Pi3 via an actively powered USB hub.

The corruption probably occured upon some powering issues. Not a full power failure, but like not enough voltage/current. When I had an extra USB device, a USB serial adapter, attached next to the USB hub made the Pi3 very unstable, resulting in crashes and kernel traces every few minutes or hours, with traces which did not make much sense. I guess that low power state might have resulted in some flipped bits.

Before attaching and after detaching the USB serial adapter the Pi3 ran stable again.

Anyway, just for some background information for what probably caused this inconsistent state.

**Comment by T-X** [comment]
And after running seaf-fsck.sh --repair without any errors, this damaged block still exists unchanged:

<pre><code>seafile@seafile:~$ sha256sum ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
[xyz]  /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
seafile@seafile:~$ stat ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
  File: /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
  Size: 4194304         Blocks: 8192       IO Block: 4096   regular file
Device: fe1dh/65053d    Inode: 198149177   Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/ seafile)   Gid: ( 1000/ seafile)
Access: 2019-06-12 12:21:46.886628204 +0000
Modify: 2019-06-09 11:46:12.709457501 +0000
Change: 2019-06-09 11:46:12.753458582 +0000
 Birth: -
seafile@seafile:~$ ~/seaf-fsck.sh --repair 8884c89b-685a-4444-9725-635b7780c905

Starting seaf-fsck, please wait ...

[06/12/19 13:42:45] fsck.c(586): Running fsck for repo 8884c89b-685a-4444-9725-635b7780c905.
[06/12/19 13:42:45] fsck.c(413): Checking file system integrity of repo foobar(8884c89b)...
[06/12/19 13:43:00] fsck.c(105): Block 201bdd34f16262991070b0908f8c548c391005c2 is damaged, remove it.
[06/12/19 13:43:00] fsck.c(182): File /somefile(aa79ad1f) is damaged, recreate an empty file.
[06/12/19 13:50:58] fsck.c(391): Update repo 8884c89b status to commit 992b7c1b.
[06/12/19 13:50:59] fsck.c(650): Fsck finished for repo 8884c89b.

seaf-fsck run done

Done.
seafile@seafile:~$ sha256sum ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
[xyz]  /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
seafile@seafile:~$ stat ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23 
  File: /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
  Size: 4194304         Blocks: 8192       IO Block: 4096   regular file
Device: fe1dh/65053d    Inode: 198149177   Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/ seafile)   Gid: ( 1000/ seafile)
Access: 2019-06-12 12:21:46.886628204 +0000
Modify: 2019-06-09 11:46:12.709457501 +0000
Change: 2019-06-09 11:46:12.753458582 +0000
 Birth: -
</code></pre>

The block mentioned in this second seaf-fsck.sh however is gone:

<pre><code>seafile@seafile:~$ find ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/ -name "*2991070b0908f8c548c391005c2*"
seafile@seafile:~$
</code></pre>

Reuploading this file to a new subdirectory in the same library, without deleting the old, 0 byte one, makes this block reappear:

<pre><code>seafile@seafile:~$ find ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/ -name "*2991070b0908f8c548c391005c2*"
/home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/20/1bdd34f16262991070b0908f8c548c391005c2
seafile@seafile:~$ sha256sum ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/20/1bdd34f16262991070b0908f8c548c391005c2 
[abc]  /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/20/1bdd34f16262991070b0908f8c548c391005c2
seafile@seafile:~$ stat ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/20/1bdd34f16262991070b0908f8c548c391005c2 
  File: /home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/20/1bdd34f16262991070b0908f8c548c391005c2
  Size: 4194304         Blocks: 8192       IO Block: 4096   regular file
Device: fe1dh/65053d    Inode: 198164313   Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/ seafile)   Gid: ( 1000/ seafile)
Access: 2019-06-12 14:10:40.630758758 +0000
Modify: 2019-06-12 14:07:52.725332230 +0000
Change: 2019-06-12 14:07:52.769331553 +0000
 Birth: -
</pre></code>

Then running seaf-fsck.sh finds another, new broken block and deletes, but does not delete the old ones that reappeared:

<pre><code>seafile@seafile:~$ ~/seaf-fsck.sh --repair 8884c89b-685a-4444-9725-635b7780c905

Starting seaf-fsck, please wait ...

[06/12/19 14:12:05] fsck.c(586): Running fsck for repo 8884c89b-685a-4444-9725-635b7780c905.
[06/12/19 14:12:05] fsck.c(413): Checking file system integrity of repo foobar(8884c89b)...
[06/12/19 14:12:21] fsck.c(105): Block 3b38f28a33281b1c7bbb110d68b6b19544ba77be is damaged, remove it.
[06/12/19 14:12:21] fsck.c(182): File /bla2/somefile(aa79ad1f) is damaged, recreate an empty file.
[06/12/19 14:19:46] fsck.c(391): Update repo 8884c89b status to commit 37dfb075.
[06/12/19 14:19:46] fsck.c(650): Fsck finished for repo 8884c89b.

seaf-fsck run done

Done.
seafile@seafile:~$ find ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/ -name "38f28a33281b1c7bbb110d68b6b19544ba77be"
seafile@seafile:~$
seafile@seafile:~$ find ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/ -name "*2991070b0908f8c548c391005c2*"
/home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/20/1bdd34f16262991070b0908f8c548c391005c2
seafile@seafile:~$ find ~/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/ -name "85f8d21805f029ab451939e9a53fa3f4ffdb23"
/home/seafile/seafile-data/storage/blocks/8884c89b-685a-4444-9725-635b7780c905/66/85f8d21805f029ab451939e9a53fa3f4ffdb23
</code></pre>

So it seems the reupload both creates a new, damaged block (which seaf-fsck.sh finds and deletes) but somehow also recovers the previously deleted block (which seaf-fsck.sh does not find).

**Comment by killing** [comment]
I think the cause is that seaf-fsck stops checking other blocks after finding and removing one damaged block. You have multiple damaged blocks in a single file. We could fix this in later version.

At the moment there is no direct fix to it. You can create a new library, copy all files to the new one.

**Comment by T-X** [comment]
Thanks! That worked.

I did a full seaf-fsck.sh --repair, then created a new library, then moved everything via the web-ui except the folder that contains/contained the broken file to this new library. Then reuploaded the previously broken file to the new library via the web-ui.

Now the download of that file works again! I did not delete the original library yet, just renamed it. So that I will hopefully still be able to reproduce the issue in there, to be able to verify any patch.

Btw., initially I think I had copied the broken file, without a previous seaf-fsck.sh after the reupload, to a new library. And then the new library seemed to have had the same issue. Does that make sense? Can the broken blocks potentially taint a new library if no seaf-fsck.sh was run before that?

---

Anyway, thanks for the amazing, quick help! Since copying and moving between libraries is so fast with Seafile this suggestion for a workaround was both very helpful and convenient.
